CRM/Import/Parser.php

```diff
@@ -646,6 +646,75 @@ protected function doPostImportActions() {
     UserJob::update(FALSE)->addWhere('id', '=', $userJob['id'])->setValues(['metadata' => $this->userJob['metadata']])->execute();
   }
 
+  /**
+   * Run import.
+   *
+   * @param \CRM_Queue_TaskContext $taskContext
+   *
+   * @param int $userJobID
+   * @param int $limit
+   * @param int $offset
+   *
+   * @return bool
+   * @throws \API_Exception
+   * @throws \CRM_Core_Exception
+   */
+  public static function runImport($taskContext, $userJobID, $limit, $offset) {
+    $parser = new CRM_Contact_Import_Parser_Contact();
+    $parser->setUserJobID($userJobID);
+    // Not sure if we still need to init....
+    $parser->init();
+    $dataSource = $parser->getDataSourceObject();
+    $dataSource->setStatuses(['new']);
+    $dataSource->setOffset($offset);
+    $dataSource->setLimit($limit);
+
+    while ($row = $dataSource->getRow()) {
+      $values = array_values($row);
+
+      try {
+        $parser->import($parser->getSubmittedValue('onDuplicate'), $values);
+      }
+      catch (CiviCRM_API3_Exception $e) {
+        // When we catch errors here we are not adding to the errors array - mostly
+        // because that will become obsolete once https://github.com/civicrm/civicrm-core/pull/23292
+        // is merged and this will replace it as the main way to handle errors (ie. update the table
+        // and move on).
+        $parser->setImportStatus((int) $values[count($values) - 1], 'ERROR', $e->getMessage());
+      }
+    }
+    $parser->doPostImportActions();
+    return TRUE;
+  }
+
+  public function queue() {
+    $dataSource = $this->getDataSourceObject();
+    $totalRowCount = $totalRows = $dataSource->getRowCount(['new']);
+    $queue = Civi::queue('user_job_' . $this->getUserJobID(), ['type' => 'Sql']);
+    $offset = 0;
+    $batchSize = 5;
+    while ($totalRows > 0) {
+      if ($totalRows < $batchSize) {
+        $batchSize = $totalRows;
+      }
+      $task = new CRM_Queue_Task(
+        [get_class($this), 'runImport'],
+        ['user_job_id' => $this->getUserJobID(), 'limit' => $batchSize, 'offset' => $offset],
+        ts('Processed %1 rows out of %2', [1 => $offset + $batchSize, 2 => $totalRowCount])
+      );
+      $queue->createItem($task);
+      $totalRows -= $batchSize;
+      $offset += $batchSize;
+    }
+    $runner = new CRM_Queue_Runner([
+      'queue' => $queue,
+      'errorMode' => CRM_Queue_Runner::ERROR_ABORT,
+      'onEndUrl' => CRM_Utils_System::url('civicrm/import/contact/summary', ['user_job_id' => $this->getUserJobID(), 'reset' => 1]),
+    ]);
+    $runner->runAllViaWeb();
```

Comment:
So in a unit-test context, `$runner->runAll()` would be more appropriate than `$runner->runAllViaWeb()`. Maybe find a conditional/variable to pick the better runner??

Reply:
@totten ok - I just looked & yes - that is the cause of the test fails - it didn't work for me in the web without that though - the specific tests ACTUALLY test the form flow - so I guess 

1) the running of it should be moved back to the form
2) the task to set up the queue should take some parameters to make it easier to offer backend as well as front end
3) the tests need to catch the premature exit exception & run the rest of the queue themselves
4) side note - in the csv api explorer (and in fact in the wmf custom import implementation - which is on it's way out) - batch size is a configuration option

![image](https://user-images.githubusercontent.com/336308/171749363-143b0639-83c6-48de-9076-38ffb8987301.png)

Reply:
Ok - so what I'm thinking to try in conjunction with your PR is 

- if a queue job already exists with a certain name (civicrm_executor or something genecidal like that) THEN offer the option to process in the background
- in this scenario we would create a queue - the same as now with a background config (error handling  'abort' ) and we would ALSO add an item to the pre-existing `civicrm_executor` to run that queue -(error handling  'discard' )
- the tricky part is I'd need 'some sort of UI' that the user could be redirected to to see what is happening because they are still 'responsible' for checking it completes at this stage.

That's the flow I figure I'll try out @totten

Reply:
Oh, the batch size option sounds like a good idea.

> the task to set up the queue should take some parameters to make it easier to offer backend as well as front end

Yeah, I think its good practice try to decouple the "build queue" and "run queue" parts. A few other examples:

* If you look at the core-upgrader, it knows how build a queue of upgrade tasks (`CRM_Upgrade_Form::buildQueue()`). Then there are a few frontends:
    * The web page `civicrm/upgrade` calls `buildQueue()` + `runAllViaWeb()`
    * The CLI `drush civicrm-upgrade-db` calls `buildQueue()` + `runAll()`
    * The CLI `cv upgrade:db` calls `buildQueue()` and its own `ConsoleQueueRunner` (*which just shows more colorful output and which respects the `-v`erbose flag*).
* If you look at ext-upgrader, it knows how to build a queue of upgrade tasks (`CRM_Extension_Upgrades::createQueue()`). Then there are a few frontends:
    * The web page `civicrm/admin/extensions/upgrade` calls `createQueue()` + `runAllViaWeb()`
    * The API `Extension.upgrade` calls `createQueue()` + `runAll()`.

Reply:
> * if a queue job already exists with a certain name (civicrm_executor or something genecidal like that) THEN offer the option to process in the background
> * in this scenario we would create a queue - the same as now with a background config (error handling 'abort' ) and we would ALSO add an item to the pre-existing civicrm_executor to run that queue -(error handling 'discard' )

OK, so for the current PR, the original scope is spot-on (ie *get it working with the foreground AJAX runner*). This is going into follow-up work, right?

FWIW, `coworker` doesn't need a `civicrm_executor` like that -- its standard behavior is to pick up on any queue which matches `runner IS NOT NULL and status = 'active'`.

I would note that `civicrm_queue` has some options that you might call *obscure tunables* (eg `lease_time`, `retry_limit`, `retry_interval`) -- things which should have safe defaults, but which sysadmins might adjust in some edge-cases. A *queue template* could serve as both configuration-element and flag-record. Ex:

```php
// Register queue-template
\Civi\Api4\Queue::create()
  ->setValues([
    'name' => 'import/template',
    'type' => 'SqlParallel',
    'runner' => 'task',
    'status' => 'template', /* ?? OR: 'is_template' => TRUE, ?? */
    'error' => 'abort',
    'lease_time' => 5*60,
    'retry_limit' => 0,
  ]),
  ->execute();
```

Then when it's time to prepare queued tasks, conditionally copy the template:

```php
if (queue('import/template') is an active template) {
  // Found template. Run in background.
  $q = Civi::queue("import/{$jobId}", ['tpl' => 'import/template', 'status' => 'draft']);
  Importer::buildQueue($q);
  $q->setStatus('active'); /* Background worker will take over */
}
else {
  // No template. Run in foreground.
  $q = Civi::queue("import/{$jobId}", ['type' => 'Sql', 'error' => 'abort']);
  Importer::buildQueue($q);
  $r = new CRM_Queue_Runner(...);
  $r->runAllViaWeb();
}
```

Reply:
@totten I'm trying to use this use case to also test & QA your PR - so yeah -  I have pulled down your PR & am trying to get it to add the background queue option in conjunction with that - but I've gonna have to have a snooze & come back to it - about 50% of my brain power won't be released until https://github.com/civicrm/civicrm-core/pull/23666 is merged & I can rebase this over it & another  45% seems to be taken up by lack of sleep so if I can chase the rabbit into the cage I'm gonna see if I can take a kip & free up some of the 45%

Reply:
& it has been merged - I rebased this